Handle "multi prompt"

wellle opened the pull request:
In a git repository my prompt looks like this:

```
git:multi-prompt nvim-editcommand: ls
```

So I set this:

``` vim
let g:editcommand_prompt = ':'
```

But because I have a `:` in the git part of my prompt, the `strip_prompt` function doesn't work. I updated it in this PR to handle this case, but I ran into a problem. To show this, I first added a commit to add some logging. Here are some examples of those logs:

```
After first commit on `~:` (no command) in `:term`:
command: "~:^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@"
part: "^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@"

After first commit on `~: ls` in `:term`:
command: "~: ls^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@"
part: "ls^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@"

After first commit on `git:multi-prompt nvim-editcommand: ls` in `:term`:
command: "git:multi-prompt nvim-editcommand: ls^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@"
part: "ulti-prompt nvim-editcommand: ls^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@"
```

Please note the `^@` characters. Those are literal `<C-@>` and I have no idea where they are from, but they are the source of the problem.

Anyway, it mostly works, except in the git repository where the wrong part gets extracted. This is what I want to fix.

In the second commit I actually changed the `strip_prompt` function to search for the right `:` by looking for a `:` followed by space or eol (like you already do in `extract_command`. Here's what I got with the same commands then:

```
After second commit on `git:multi-prompt nvim-editcommand: ls` in `:term`:
command: "git:multi-prompt nvim-editcommand: ls^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@"
part: "ls^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@"

After second commit on `~: ls` in `:term`:
command: "~: ls^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@"
part: "ls^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@"

After second commit on `~:` (no command) in `:term`:
command: "~:^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@"
part: "~:^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@"
```

You see that it now works both in git and non-git repository directories as long as there is a command after the prompt. If there is no command, those pesky `^@` characters make the eol not match.

Could you see (by using the first commit only) if you have those `^@` characters as well? Do you have an idea where they might be coming from?

I could fix it by updating the regex to ignore those characters before eol, but it would be better if we could understand what's actually going on here.

Once this is resolved I can clean up the PR by removing the logging again.

wellle committed:
Use match() instead of stridx to find prompt_idx

brettanomyces commented:
I believe that the `^@` characters come from newline characters that do not get handled properly, although I'm yet identify why/where this mishandling happens. The reason there are so many of them is that in `NORMAL` mode Neovim treats all the visible space after the prompt as empty lines, containing only a newline character. 

I'm at work at the moment but I'll look into it tonight.

wellle commented:
Thanks!

brettanomyces commented:
So as I understand it. the `^@` are newline/<c-j> being displayed instead of an actual newline.  

`:echom` displays them as `^@` but `:echo` displays the actual newline. 

To see then run the following commands in a terminal buffer:

```
:echom join(getline(1, '$'), "\n")
```

vs.

```
:echo join(getline(1, '$'), "\n")
```

The newlines are there because Neovim treats space after the commandline in a terminal buffer as empty lines.

The newlines are removed by the format_command() function after the command has been put into  the scratch buffer/temporary file.

So I think we're good to clean up and merge your fix to handle cursor characters before the actual cursor.

Some comments:
- I really like how you use `\zs`, I don't think I would have though of that.
- `\zs` by itself is sufficient, we don't need the `.*`. 
- we   could make `l:space_or_eol` script local (`s:`) and define at the start of the file, after the global variable instead of defining it in two separate places.

wellle committed:
Add linebreak to s:space_or_eol, make it script local

wellle commented:
That explains it :+1:
- I removed `.*` after `\zs` in the `match()` commit.
- I added a commit to add `\n` to `s:space_or_eol` to deal with `^@` and made it script local.
- I removed the commit that added the logging.

I tested all cases again and it all seems to be working now. Let me know if there's anything left to do.

brettanomyces commented:
Looks good to me. Thanks very much!
